docs/spec/spec.md

```diff
@@ -0,0 +1,1380 @@
+}
+service /graphql on securedEP {
+    resource function get greeting() returns string {
+        return "Hello, World!";
+    }
+}
+```
+
+##### 8.1.1.4 OAuth2
+
+A GraphQL service can be secured using [OAuth2](https://github.com/ballerina-platform/module-ballerina-oauth2/blob/master/docs/spec/spec.md) and by enforcing authorization optionally.
+
+When configured, it validates the OAuth2 token sent in the `Authorization` header in the HTTP request that contains the GraphQL document. This calls the configured OAuth2 introspection endpoint to validate.
+
+###### Example: Declarative OAuth2
+
+```ballerina
+@graphql:ServiceConfig {
+    auth: [
+        {
+            oauth2IntrospectionConfig: {
+                url: "https://localhost:9445/oauth2/introspect",
+                tokenTypeHint: "access_token",
+                scopeKey: "scp",
+                clientConfig: {
+                    customHeaders: {"Authorization": "Basic YWRtaW46YWRtaW4="},
+                    secureSocket: {
+                        cert: "../resource/path/to/public.crt"
+                    }
+                }
+            },
+            scopes: ["admin"]
+        }
+    ]
+}
+service /graphql on securedEP {
+    resource function get greeting() returns string {
+        return "Hello, World!";
+    }
+}
+```
+
+#### 8.1.2 Imperative Approach
+
+This is also known as the code-driven approach, which is used for advanced use cases, where users need to be worried more about how authentication and authorization work and need to have further customizations. The user has full control of the code-driven approach. The handler creation and authentication/authorization calls are made by the user at the business logic layer.
+
+The [`graphql:Context`](#6-context) object and the [`contextInit`](#713-context-initializer-function) method can be used to achieve this.
+
+##### 8.1.2.1 Basic Auth - File User Store
+
+A file user store can be used to validate the `Authorization` header in the HTTP request that contains the GraphQL document.
+
+###### Example: Imperative Basic Auth with File User Store
+
+```ballerina
+graphql:FileUserStoreConfig config = {};
+http:ListenerFileUserStoreBasicAuthHandler handler = new (config);
+
+isolated function contextInit(http:RequestContext reqCtx, http:Request request) returns graphql:Context|error {
+    string authorization = check request.getHeader("Authorization");
+    graphql:Context context = new;
+    context.set("Authorization", authorization);
+    return context;
+}
+
+@graphql:ServiceConfig {
+    contextInit: contextInit
+}
+service on new graphql:Listener(4000) {
+    resource function get greeting(graphql:Context context) returns string|error {
+        value:Cloneable|isolated object {} authorization = check context.get("Authorization");
+        if authorization !is string {
+            return error("Failed to authorize");
+        }
+        auth:UserDetails|http:Unauthorized authn = handler.authenticate(Authorization);
```

Comment:
```suggestion
        auth:UserDetails|http:Unauthorized authn = handler.authenticate(authorization);
```
Also, please check whether we need to cast to `string` here.

Reply:
We don't need to cast here since we are returning the value in the above type guard.